ap2000   add shift register for c000 - fixes the missing 9th pin line

goldnchild opened the pull request.

cuavas commented on src/devices/machine/e05a30.cpp:
```diff
@@ -167,6 +168,13 @@ void e05a30_device::write(offs_t offset, uint8_t data)
+	case 0x1:
+	case 0x2:
+	case 0x3:
+		m_c000_shift_register &= ~(0xff << ((3-offset)*8));
+		m_c000_shift_register |= (data << ((3-offset)*8));
```
You’re relying on integer promotion rules here – please make it explicit and cast `0xff` and `data` to the desired type.

cuavas commented on src/devices/machine/e05a30.cpp:
```diff
@@ -167,6 +168,13 @@ void e05a30_device::write(offs_t offset, uint8_t data)
 	LOG("%s: e05a30_w([0xC0%02x]): %02x\n", machine().describe_context(), offset, data);
 
 	switch (offset) {
+	case 0x0:
+	case 0x1:
+	case 0x2:
+	case 0x3:
```
The rest of this `switch` statement has two-digit hex case constants.  Please be consistent with it.

cuavas commented on src/devices/machine/e05a30.cpp:
```diff
@@ -197,6 +205,10 @@ uint8_t e05a30_device::read(offs_t offset)
 	LOG("%s: e05a30_r([0xC0%02x]): ", machine().describe_context(), offset);
 
 	switch (offset) {
+	case 0x0:
```
Please be consistent with the rest of this `switch` statement, too.

cuavas commented on src/devices/machine/e05a30.h:
```diff
@@ -76,6 +76,7 @@ class e05a30_device : public device_t
 	uint8_t m_centronics_strobe;
 	uint8_t m_centronics_data_latch;
 	uint8_t m_centronics_data_latched;
+	int m_c000_shift_register;
```
Please use an explicitly-sized type for this.

cuavas commented on src/devices/machine/e05a30.cpp:
```diff
@@ -197,6 +205,10 @@ uint8_t e05a30_device::read(offs_t offset)
 	switch (offset) {
+	case 0x0:
+		m_c000_shift_register <<= 1;
+		result = (m_c000_shift_register & 0x1000000) ? 0x80 : 0x0;
```
IMO this would be easier to read as `BIT(m_c000_shift_register, 24) << 7` – it makes it obvious you’re trying extract bit 24 of the value into bit 7 of the result.

cuavas commented on src/devices/machine/e05a30.cpp:
```diff
@@ -197,6 +205,10 @@ uint8_t e05a30_device::read(offs_t offset)
 
 	switch (offset) {
+	case 0x0:
+		m_c000_shift_register <<= 1;
```
This needs to be guarded for debugger reads, etc.

cuavas reviewed (changes requested):
Is the shift register really 32 bits?  It doesn’t really make sense that way because the top eight bits are completely inaccessible.  Writing to offset 0 has no visible effect, since you can only access the bit shifted out of the low 24 bits.

The write would make more sense as (without handling 0x00):
```
    case 0x01:
    case 0x02:
    case 0x03:
        m_c000_shift_register &= ~(uint32_t(0xff) << ((3 - offset) * 8));
        m_c000_shift_register |= (uint32_t(data) << ((3 - offset) * 8));
        break;

And read would make more sense as (remembering the debugger and other things can trigger reads):
```
    case 0x00:
        result = BIT(m_c000_shift_register, 23) << 7;
        if (!machine().side_effects_disabled())
            m_c000_shift_register = (m_c000_shift_register << 1) & 0xffffff;
        break;
```

cuavas commented on src/devices/machine/e05a30.cpp:
```diff
@@ -197,6 +207,13 @@ uint8_t e05a30_device::read(offs_t offset)
 	switch (offset) {
+	case 0x00:
+		if (!machine().side_effects_disabled()) {
+			m_c000_read = BIT(m_c000_shift_register, 23) << 7;
```
You don’t need this extra `m_c000_read` member.  Extracting a bit from the shift register doesn’t have side effects.  The extra member is redundant can just something that can go wrong.

cuavas commented on src/devices/machine/e05a30.cpp:
```diff
@@ -167,6 +168,15 @@ void e05a30_device::write(offs_t offset, uint8_t data)
 	LOG("%s: e05a30_w([0xC0%02x]): %02x\n", machine().describe_context(), offset, data);
 
 	switch (offset) {
+	case 0x00:
+		m_c000_shift_register = 0;
+		m_c000_read = 0;
```
This is a change in behaviour – you’re now clearing the shift register on a write to offset zero.  Is that correct?  Also, you’re apparently missing a `break` statement here.

goldnchild commented:
Ok, I've incorporated your suggested edits.

I don't know if it's correct, but I made writing to c000 clear the shift register.  In the ap2000 code, they seem to write 0 to c000 so I'm guessing it's to clear the register.

I added a separate variable to hold the high bit that gets shifted out of the shift register.  That way, any subsequent reads by the debugger will show the last value that was read out.

It's interesting to note that there's a really good information on the e05a03 that tells all about the 24 bit shift register (LX-800 Technical manual) but the (LX-810/850 Technical manual) which covers the e05a30 doesn't mention a shift register at all.

cuavas commented:
> I don't know if it's correct, but I made writing to c000 clear the shift register. In the ap2000 code, they seem to write 0 to c000 so I'm guessing it's to clear the register.

Well at least comment that it’s a guess then.

> I added a separate variable to hold the high bit that gets shifted out of the shift register. That way, any subsequent reads by the debugger will show the last value that was read out.

This is unintuitive – you want the debugger to show the value that would be seen if the register were read now.

goldnchild commented:
Ok, c000 read now gets bit 23, and doesn't use another variable.

And documented c000 write clearing the shift register as a guess.

cuavas commented on src/devices/machine/e05a30.cpp:
```diff
@@ -167,6 +168,21 @@ void e05a30_device::write(offs_t offset, uint8_t data)
+		 * is incomplete. There are instances of writing 0 to c000, so it is a guess
+		 * that writing to c000 will clear the shift register. */
+	case 0x00:
+		m_c000_shift_register = 0;
```
You’re still missing a break here.  It’s falling through and setting the non-existent bits 24-31 of the shift register.

goldnchild commented:
sorry, I missed that break
